### PR TITLE
Restore jsonPath and fix it

### PR DESCRIPTION
**Pull request by benmoriceau**
## What

This restore the Json traversal library. A bug was introduce in the Json path library, the PR fix it.

In a json schema we can define an enum without specifying a "type" attribute. It wasn't handle in the previous implemantation. We now return a right type in the `getType` method and process it the same way than the an integer/boolean/string type.

## File to look at

There are 2 commits to look at:
- one for the revert, it has already been reviewed
- The other one for a fix, The main fix logic is in line 224  of the files `JsonSchemas` and it got tested with the addition of the `enum` folder of the schema and configuration test suite. The files are actual configuration which has caused an error.

### CODE REVIEW

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java [hunk 1]
```diff
@@ -220,6 +221,9 @@ public static List<String> getType(final JsonNode jsonNode) {
         return List.of(jsonNode.get(JSON_SCHEMA_TYPE_KEY).asText());
       }
     }
+    if (jsonNode.has(ENUM_TYPE)) {
+      return List.of(ENUM_TYPE);
```
Comment (cgardens) [inline airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java:225]: Do we need to return a new type? Can we say the enum type is always a string instead of adding a new type? i.e. let's return `List.of(STRING_TYPE)`. Does JSONSchema say anything about this? I thought the enum field was a modifier for string types only but I could be wrong about what the specification says.

Also, is just having the enum field considered valid JSONSchema? Or is it just something that _we_ do? If the latter, let's add a comment on line 224 calling that out so that it is clear. I'm fine with us being a little more permissive, but I want to be clear about whether we are breaking the normal convention or not.

In the future, it would be good to add to the standard tests for a connector a validation that the specs it uses match the subset of JSONSchema functionality that we support (certainly out of scope for this issue).

**Review by cgardens — CHANGES_REQUESTED** [review]
nice catch! Let one open question below. once that is resolved we will be good to go.

File: airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java (hunk 1 above)
Comment (cgardens) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java:225]: created this issue for the future work i proposed above: https://github.com/airbytehq/airbyte/issues/12330

Comment (benmoriceau) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java:225]: I have seen a stack-overflow example with integer type for the enum while searching for the best apporach to this fix https://stackoverflow.com/questions/30924271/correct-way-to-define-array-of-enums-in-json-schema. So I wanted to introduce specific type because nothing tell me in the specs that a given type need to be enforce.

Other source don't explicit the type https://json-schema.org/understanding-json-schema/reference/generic.html#enumerated-values

Comment (cgardens) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java:225]: > [6.1.2.](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.1.2) [enum](https://json-schema.org/draft/2020-12/json-schema-validation.html#enum)
> The value of this keyword MUST be an array. This array SHOULD have at least one element. Elements in the array SHOULD be unique.
> 
> An instance validates successfully against this keyword if its value is equal to one of the elements in this keyword's array value.
> 
> Elements in the array might be of any type, including null.
-- https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.1.2

huh. you are right. enums are not restricted to strings.

I think the key thing here is that `enum` is _not_ a JSONSchema type. It is a modifier field. 
> [6.1.1.](https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.1.1) type
> The value of this keyword MUST be either a string or an array. If it is an array, elements of the array MUST be strings and MUST be unique.
>
> String values MUST be one of the six primitive types ("null", "boolean", "object", "array", "number", or "string"), or "integer" which matches any number with a zero fractional part.
>
> An instance validates if and only if the instance is in any of the sets listed for this keyword.

-- https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.6.1.1

My reading of this is that it is invalid for a JSONSchema object to not have a `type` field set. I think right now we permit it and assume if it is missing it is an `object`. That permissiveness bites us now, because really the assumption that would make sense if no type is set but `enum` is set is probably string, but this starts getting complicated.

I don't think we should introduce `enum` as a new type. It isn't real JSONSchema, but someone who is not very familiar with is not going to know that. I think in the short term I would suggest: 1) if enum but type not set, assume string. 2) fix the invalid specs (though this may be very hard to do).

Comment (cgardens) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java:225]: Ha! I'm not crazy. Other people are confused by this too: https://github.com/json-schema/json-schema/issues/172. But also means my previous statement is wrong. If `type` is not set, you can make no assumptions about the type of the object. It can be anything!

Okay, based on that I do think my suggestions from the previous section holds. I think for `specs`, _we_ should not allow the `type` to not be set. There is no reason to saddle ourselves with that ambiguity. That said, there is not a "free" path to that without giving connector developers a tool to validate their schemas. Thus, a shim for now that just assumes that if `type` is not set and `enum` is set that we are dealing with a string seems like an okay hack. And then we can remove it once we provide adequate tooling for connector developers in this future issue https://github.com/airbytehq/airbyte/issues/12330. WDYT?

Comment (benmoriceau) [reply airbyte-commons/src/main/java/io/airbyte/commons/json/JsonSchemas.java]: I think it is the right way to go, even if the types are not specify. In this case we want the same behavior for string and enum since both are non-traversable and thus are a stop condition in the recursion.

We can then work on fixing the schema.

**Review by cgardens — APPROVED** [review]
nice! LGTM.

**Comment by benmoriceau** [comment]
@cgardens FYI: I have added `SpecFormatTest` in order to test the new java validation over all the specs. I have update 2 specs that were having errors. I will try to find who is the best for reviewing it in the connector team.

**Comment by benmoriceau** [comment]
@girarda @evantahler I have added you both to this review because I modified 2 source specs that were incorrect based on our implementation of the JSON-Schema traversal. The only file that should be a concern to you are the one in the `airbyte-integrations/connector` folder.

File: airbyte-config/init/src/main/resources/seed/source_specs.yaml [hunk 2]
```diff
@@ -6,7 +6,7 @@
     documentationUrl: "https://docs.airbyte.io/integrations/sources/airtable"
     connectionSpecification:
       $schema: "http://json-schema.org/draft-07/schema#"
-      title: "Airtable Source Spec"
+      title: "Airtadfgdfgble Source Spec"
```
Comment (girarda) [inline airbyte-config/init/src/main/resources/seed/source_specs.yaml:9]: is this change intentional?

Comment (benmoriceau) [reply airbyte-config/init/src/main/resources/seed/source_specs.yaml:9]: no, will update thanks for spotting it.

**Review by girarda — APPROVED** [review]
